torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -716,11 +715,20 @@ def reset_sharded_param(self):
             padded_local_tensor = local_tensor.new_zeros(padded_sharded_size)
             padded_local_tensor[: local_tensor.size(0)].copy_(local_tensor)
             local_tensor = padded_local_tensor
+        if self.offload_to_cpu and not local_tensor.is_cpu:
+            local_tensor = local_tensor.cpu()
         if self.pin_memory and not local_tensor.is_pinned():
             local_tensor = local_tensor.cpu().pin_memory()
         self._sharded_param_data = local_tensor.view(-1)
-        assert isinstance(self.sharded_param, DTensor)  # mypy
-        self.sharded_param._local_tensor = local_tensor[: self.sharded_size[0]]
+        if not isinstance(new_param, DTensor):
+            raise ValueError(
+                f"expect the new parameter to be DTensor but got {type(new_param)}"
+            )
+        self.sharded_param = nn.Parameter(
+            self.to_sharded_dtensor(local_tensor[: self.sharded_size[0]]),
+            requires_grad=new_param.requires_grad,
+        )
+        self._setattr_on_modules(self.sharded_param)
```

Comment:
this won't work because optimizer.param_groups refers to a different `model.parameters()`